document/core/appendix/algorithm.rst

```diff
@@ -43,38 +46,46 @@ Value types are representable as a set of enumerations.
    func is_ref(t : val_type) : bool =
      return not (is_num t || is_vec t) || t = Bot
 
-Equivalence and subtyping checks can be defined on these types.
+Similarly, :ref:`defined types <syntax-deftype>` :code:`def_type` can be represented:
 
 .. code-block:: pseudo
 
-   func eq_def(n1, n2) =
-     ...  // check that both type definitions are equivalent (TODO)
-
-   func matches_null(null1 : bool, null2 : bool) : bool =
-     return null1 = null2 || null2
-
-   func matches_heap(t1 : heap_type, t2 : heap_type) : bool =
-     switch (t1, t2)
-       case (Def(n1), Def(n2))
-         return eq_def(n1, n2)
-       case (Def(_), Func)
-         return true
-       case (Bot, _)
-         return true
-       case (_, _)
-         return t1 = t2
-
-   func matches_ref(t1 : ref_type, t2 : ref_type) : bool =
-     return matches_heap(t1.heap, t2.heap) && matches_null(t1.null, t2.null)
-
-   func matches(t1 : val_type, t2 : val_type) : bool =
-     switch (t1, t2)
-       case (Ref(_), Ref(_))
-         return matches_ref(t1, t2)
-       case (Bot, _)
-         return true
-       case (_, _)
-         return t1 = t2
+   type packed_type = I8 | I16
+   type field_type = Field(val : val_type | packed_type, mut : bool)
+
+   type struct_type = Struct(fields : list(field_type))
+   type array_type = Array(fields : field_type)
+   type func_type = Func(params : list(val_type), results : list(val_type))
+   type comp_type = struct_type | array_type | func_type
+
+   type sub_type = Sub(super : list(def_type), body : comp_type, final : bool)
+   type rec_type = Rec(types : list(sub_type))
+
+   type def_type = Def(rec : rec_type, proj : int32)
+
+   func unpack_field(t : field_type) : val_type =
+     if (it = I8 || t = I16) return I32
+     return t
+
+   func expand_def(t : def_type) : comp_type =
+     return t.rec.types[t.proj].body
+
+These representations assume that all types have been :ref:`closed <type-closed>` by :ref:`substituting <aux-subst>` all :ref:`type indices <syntax-typeidx>` (in :ref:`concrete heap types <syntax-heaptype>` and in :ref:`sub types <syntax-subtype>`) with their respective :ref:`defined types <syntax-deftype>`.
+This includes *recursive* references to enclosing :ref:`defined types <syntax-deftype>`,
+such that type representations form graphs and may be *cyclic* for :ref:`recursive types <syntax-rectype>`.
+
+We assume that all types have been *canonicalized*, such that equality on two type representations holds if and only if they are in fact :ref:`equivalent <equiv>`. making it a constant-time check.
```

Comment:
Do we plan to sketch an algorithm for this?

Reply:
For canonicalisation? No, that's a standard algorithm on trees.

I started writing out the subtyping algorithm, but concluded that it wasn't interesting enough either. After all, the point of this appendix is to show a sketch of a linear validation algorithm, because that can be written quite differently from the declarative rules. That's not the case for other parts of the semantics.